File: app/src/main/java/com/remotearthsolutions/expensetracker/databaseutils/models/dtos/CategoryExpense.java
package com.remotearthsolutions.expensetracker.databaseutils.models.dtos;

import androidx.room.Ignore;
import com.remotearthsolutions.expensetracker.databaseutils.models.AccountModel;
import com.remotearthsolutions.expensetracker.databaseutils.models.CategoryModel;
import com.remotearthsolutions.expensetracker.utils.DateTimeUtils;
import org.parceler.Parcel;

@Parcel
public class CategoryExpense {

    private int expense_id;
    private int category_id;
    private String category_name;
    private String category_icon;
    private double total_amount;
    private long datetime;
    private int account_id;
    private String account_name;
    private String account_icon;
    private String note;

    @Ignore
    public boolean isHeader = false;

    public CategoryExpense(int category_id, String category_name, String category_icon, double total_amount, long datetime) {
        this.category_id = category_id;
        this.category_name = category_name;
        this.category_icon = category_icon;
        this.total_amount = total_amount;
        this.datetime = datetime;
    }

    public CategoryExpense(int expense_id, int category_id, String category_name, String category_icon, double total_amount, long datetime,
                           int account_id, String account_name, String account_icon, String note) {
        this.expense_id = expense_id;
        this.category_id = category_id;
        this.category_name = category_name;
        this.category_icon = category_icon;
        this.total_amount = total_amount;
        this.datetime = datetime;
        this.account_id = account_id;
        this.account_name = account_name;
        this.account_icon = account_icon;
        this.note = note;
    }

    public CategoryExpense() {
    }

    public int getExpense_id() {
        return expense_id;
    }

    public void setExpense_id(int expense_id) {
        this.expense_id = expense_id;
    }

    public int getCategory_id() {
        return category_id;
    }

    public void setCategory_id(int category_id) {
        this.category_id = category_id;
    }

    public String getCategory_name() {
        return category_name;
    }

    public void setCategory_name(String category_name) {
        this.category_name = category_name;
    }

    public String getCategory_icon() {
        return category_icon;
    }

    public void setCategory_icon(String category_icon) {
        this.category_icon = category_icon;
    }

    public double getTotal_amount() {
        return total_amount;
    }

    public void setTotal_amount(double total_amount) {
        this.total_amount = total_amount;
    }

    public long getDatetime() {
        return datetime;
    }

    public void setDatetime(long datetime) {
        this.datetime = datetime;
    }

    public int getAccount_id() {
        return account_id;
    }

    public void setAccount_id(int account_id) {
        this.account_id = account_id;
    }

    public String getAccount_name() {
        return account_name;
    }

    public void setAccount_name(String account_name) {
        this.account_name = account_name;
    }

    public String getAccount_icon() {
        return account_icon;
    }

    public void setAccount_icon(String account_icon) {
        this.account_icon = account_icon;
    }

    public String getNote() {
        return note;
    }

    public void setNote(String note) {
        this.note = note;
    }

    @Override
    public String toString() {
        return DateTimeUtils.getDate(datetime, DateTimeUtils.dd_MM_yyyy) + ", " +
                category_name + ", " +
                total_amount + ", " +
                account_name + ", " +
                note + "\n";
    }

    public void setCategory(CategoryModel category) {
        this.category_id = category.getId();
        this.category_name = category.getName();
        this.category_icon = category.getIcon();
    }

    public void setAccount(AccountModel account) {
        this.account_id = account.getId();
        this.account_name = account.getName();
        this.account_icon = account.getIcon();
    }
}

File: app/src/main/java/com/remotearthsolutions/expensetracker/databaseutils/models/dtos/CategoryExpense.kt
package com.remotearthsolutions.expensetracker.databaseutils.models.dtos

import androidx.room.ColumnInfo
import androidx.room.Ignore
import com.remotearthsolutions.expensetracker.databaseutils.models.AccountModel
import com.remotearthsolutions.expensetracker.databaseutils.models.CategoryModel
import com.remotearthsolutions.expensetracker.utils.DateTimeUtils
import org.parceler.Parcel

@Parcel(Parcel.Serialization.BEAN)
class CategoryExpense {
    @ColumnInfo(name = "expense_id")
    var expenseId = 0
    @ColumnInfo(name = "category_id")
    var categoryId = 0
    @ColumnInfo(name = "category_name")
    var categoryName: String? = null
    @ColumnInfo(name = "category_icon")
    var categoryIcon: String? = null
    @ColumnInfo(name = "total_amount")
    var totalAmount = 0.0
    @ColumnInfo(name = "datetime")
    var datetime: Long = 0
    @ColumnInfo(name = "account_id")
    var accountId = 0
    @ColumnInfo(name = "account_name")
    var accountName: String? = null
    @ColumnInfo(name = "account_icon")
    var accountIcon: String? = null
    @ColumnInfo(name = "note")
    var note: String? = null
    @Ignore
    var isHeader = false

    constructor(
        category_id: Int,
        category_name: String?,
        category_icon: String?,
        total_amount: Double,
        datetime: Long
    ) {
        categoryId = category_id
        categoryName = category_name
        categoryIcon = category_icon
        totalAmount = total_amount
        this.datetime = datetime
    }

    constructor(
        expense_id: Int,
        category_id: Int,
        category_name: String?,
        category_icon: String?,
        total_amount: Double,
        datetime: Long,
        account_id: Int,
        account_name: String?,
        account_icon: String?,
        note: String?
    ) {
        expenseId = expense_id
        categoryId = category_id
        categoryName = category_name
        categoryIcon = category_icon
        totalAmount = total_amount
        this.datetime = datetime
        accountId = account_id
        accountName = account_name
        accountIcon = account_icon
        this.note = note
    }

    constructor() {}

    override fun toString(): String {
        return DateTimeUtils.getDate(datetime, DateTimeUtils.dd_MM_yyyy_h_mm).toString() + ", " +
                categoryName + ", " +
                totalAmount + ", " +
                accountName + ", " +
                note + "\n"
    }

    fun setCategory(category: CategoryModel) {
        categoryId = category.id
        categoryName = category.name
        categoryIcon = category.icon
    }

    fun setAccount(account: AccountModel) {
        accountId = account.id
        accountName = account.name
        accountIcon = account.icon
    }

    fun copy() : CategoryExpense{
        val categoryExpense = CategoryExpense()
        categoryExpense.expenseId = this.expenseId
        categoryExpense.categoryId = this.categoryId
        categoryExpense.categoryName = this.categoryName
        categoryExpense.categoryIcon = this.categoryIcon
        categoryExpense.totalAmount = this.totalAmount
        categoryExpense.datetime = this.datetime
        categoryExpense.accountId = this.accountId
        categoryExpense.accountName = this.accountName
        categoryExpense.accountIcon = this.accountIcon
        categoryExpense.note = this.note
        categoryExpense.isHeader = this.isHeader
        return categoryExpense
    }
}